File: challenges/Algorithmic/Random Name Picker/rnp.py
"""rnp.py - Random name picker with modern CLI & features.

Enhancements over the original script:
* Dataclass configuration object
* argparse CLI with JSON output option
* Optional weighted selection via "name,weight" lines
* Support picking with or without replacement
* Reproducibility via --seed
* Automatic default file creation
* Clean separation of parsing, selection, and I/O for testability

File Format
===========
Each non-empty line is either:
  Name
  Name,weight   (weight must be a positive number)

Examples:
  Alice,3
  Bob,1
  Charlie

CLI Examples
------------
Pick 3 unique names:
    python rnp.py -c 3

Allow repeats (with replacement):
    python rnp.py -c 5 --with-replacement

Weighted selection (if weights in file) and JSON output:
    python rnp.py -c 4 --json

Deterministic run:
    python rnp.py -c 4 --seed 42
"""

from __future__ import annotations

import argparse
import json
import os
import random
import sys
from dataclasses import dataclass
from typing import Iterable, List, Optional, Sequence, Tuple


# ---------------------------- Data Model ---------------------------- #
@dataclass(slots=True)
class NamePickerConfig:
    path: str = "names.txt"
    count: int = 1
    with_replacement: bool = False
    json_output: bool = False
    seed: Optional[int] = None
    create_default: bool = True

    def validate(self, total_names: int) -> None:
        if self.count <= 0:
            raise ValueError("Count must be positive")
        if not self.with_replacement and self.count > total_names:
            raise ValueError(
                f"Cannot pick {self.count} unique names from {total_names}"
            )


DEFAULT_NAMES = [
    "Alice",
    "Bob",
    "Charlie",
    "David",
    "Eve",
    "Frank",
    "Grace",
    "Heidi",
    "Ivan",
    "Judy",
]


def ensure_default_file(path: str) -> None:
    if os.path.exists(path):
        return
    try:
        with open(path, "w", encoding="utf-8") as f:
            for name in DEFAULT_NAMES:
                f.write(name + "\n")
        print(f"Created default name file at '{path}'")
    except OSError as e:  # pragma: no cover - unlikely
        print(f"Error creating default file: {e}", file=sys.stderr)
        sys.exit(1)


def parse_names_file(path: str) -> Tuple[List[str], Optional[List[float]]]:
    """Parse names (and optional weights) from file.

    Returns (names, weights) where weights may be None if no weights present.
    If any line has a weight, missing weights default to 1.0.
    """
    names: List[str] = []
    weights: List[float] = []
    saw_weight = False
    with open(path, "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            if "," in line:
                name, w_str = [part.strip() for part in line.split(",", 1)]
                try:
                    w = float(w_str)
                    if w <= 0:
                        raise ValueError
                except ValueError:
                    print(
                        f"Warning: invalid weight '{w_str}' for name '{name}' – skipping line",
                        file=sys.stderr,
                    )
                    continue
                names.append(name)
                weights.append(w)
                saw_weight = True
            else:
                names.append(line)
                weights.append(1.0)
    if not names:
        raise ValueError("Name file is empty")
    if not saw_weight:
        return names, None
    return names, weights


def pick_names(
    names: Sequence[str],
    count: int,
    *,
    with_replacement: bool,
    weights: Optional[Sequence[float]],
    rng: random.Random,
) -> List[str]:
    if with_replacement:
        # random.choices uses weights for replacement scenario
        return rng.choices(names, weights=weights, k=count)
    # Without replacement: use sample; weights necessitate manual approach.
    if weights is None:
        return rng.sample(list(names), count)
    # Weighted without replacement: perform sequential weighted selection.
    chosen: List[str] = []
    names_mut = list(names)
    weights_mut = list(weights)
    for _ in range(count):
        total = sum(weights_mut)
        r = rng.random() * total
        acc = 0.0
        for i, w in enumerate(weights_mut):
            acc += w
            if r <= acc:
                chosen.append(names_mut.pop(i))
                weights_mut.pop(i)
                break
    return chosen


def build_arg_parser() -> argparse.ArgumentParser:
    p = argparse.ArgumentParser(
        description="Random name picker with optional weighting"
    )
    p.add_argument("-c", "--count", type=int, default=1, help="Number of names to pick")
    p.add_argument(
        "-f", "--file", dest="path", default="names.txt", help="Path to names file"
    )
    p.add_argument(
        "--with-replacement",
        action="store_true",
        help="Allow picking the same name multiple times",
    )
    p.add_argument(
        "--json", action="store_true", help="Output JSON instead of plain text"
    )
    p.add_argument("--seed", type=int, help="Seed RNG for reproducibility")
    p.add_argument(
        "--no-create-default",
        action="store_true",
        help="Do not auto-create default file if missing",
    )
    return p


def main(argv: Optional[Iterable[str]] = None) -> int:
    parser = build_arg_parser()
    args = parser.parse_args(list(argv) if argv is not None else None)
    cfg = NamePickerConfig(
        path=args.path,
        count=args.count,
        with_replacement=args.with_replacement,
        json_output=args.json,
        seed=args.seed,
        create_default=not args.no_create_default,
    )

    if cfg.create_default:
        ensure_default_file(cfg.path)
    if not os.path.exists(cfg.path):
        print(f"Error: Name file '{cfg.path}' not found", file=sys.stderr)
        return 1

    try:
        names, weights = parse_names_file(cfg.path)
        cfg.validate(len(names))
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        return 1

    rng = random.Random(cfg.seed)
    picked = pick_names(
        names,
        cfg.count,
        with_replacement=cfg.with_replacement,
        weights=weights,
        rng=rng,
    )

    if cfg.json_output:
        payload = {
            "picked": picked,
            "count": cfg.count,
            "with_replacement": cfg.with_replacement,
            "seed": cfg.seed,
            "total_names": len(names),
            "weights_used": weights is not None,
        }
        print(json.dumps(payload, indent=2))
    else:
        print(f"Picked {len(picked)} name(s):")
        for name in picked:
            print(f"  - {name}")
        if cfg.seed is not None:
            print(f"(Deterministic with seed {cfg.seed})")
    return 0


if __name__ == "__main__":  # pragma: no cover
    raise SystemExit(main())

File: challenges/Algorithmic/Random Name Picker/RNP.hs
{-# LANGUAGE RecordWildCards #-}
-- |
-- Module      : RNP
-- Description : Random name picker executable written in Haskell.
--
-- Mirrors the behaviour of the Python implementation (rnp.py) by supporting
-- weighted draws, deterministic seeding, optional JSON output, and a simple
-- statistics block that downstream visualisers can consume.
-- The script intentionally avoids external dependencies so it can be executed
-- with "runghc" on a stock GHC installation.
module Main (main) where

import Control.Exception (IOException, try)
import Control.Monad (foldM, when)
import Data.Char (isSpace)
import Data.List (intercalate)
import Data.Maybe (fromMaybe)
import System.Environment (getArgs)
import System.Exit (exitFailure)
import System.IO (hPutStrLn, stderr)
import System.Random (StdGen, mkStdGen, newStdGen, randomR)

-- | Runtime configuration parsed from CLI arguments.
data Config = Config
  { cfgPath :: FilePath
  , cfgCount :: Int
  , cfgWithReplacement :: Bool
  , cfgJSON :: Bool
  , cfgSeed :: Maybe Int
  , cfgStats :: Bool
  }
  deriving (Show, Eq)

defaultConfig :: Config
defaultConfig =
  Config
    { cfgPath = "names.txt"
    , cfgCount = 1
    , cfgWithReplacement = False
    , cfgJSON = False
    , cfgSeed = Nothing
    , cfgStats = False
    }

-- | Parsed input data. The weight vector is present when at least one
-- weight was found in the file. When absent, all names have an implicit
-- weight of 1.0.
data NameData = NameData
  { ndNames :: [String]
  , ndWeights :: Maybe [Double]
  }
  deriving (Show, Eq)

-- | Known default names for quick bootstrapping.
defaultNames :: [String]
defaultNames =
  [ "Alice"
  , "Bob"
  , "Charlie"
  , "David"
  , "Eve"
  , "Frank"
  , "Grace"
  , "Heidi"
  , "Ivan"
  , "Judy"
  ]

-- | Command line usage text shown for -h/--help or parsing failures.
usage :: String
usage =
  unlines
    [ "Random Name Picker (Haskell)"
    , ""
    , "Usage: runghc RNP.hs [options]"
    , ""
    , "Options:"
    , "  -c, --count N            Number of names to pick (default 1)"
    , "  -f, --file PATH          Path to names file (default names.txt)"
    , "      --with-replacement  Allow selecting the same name multiple times"
    , "      --json              Emit JSON output"
    , "      --stats             Include visualiser-ready statistics"
    , "      --seed N            Seed for deterministic runs"
    , "  -h, --help              Show this help text"
    ]

main :: IO ()
main = do
  args <- getArgs
  case parseArgs args of
    Left err -> do
      hPutStrLn stderr err
      putStrLn usage
      exitFailure
    Right cfg -> runWithConfig cfg

-- | Parse CLI arguments using a hand-rolled recursive descent. The parser
-- keeps the dependency footprint minimal and mirrors the Python CLI flags.
parseArgs :: [String] -> Either String Config
parseArgs = go defaultConfig
 where
  go cfg [] = Right cfg
  go cfg (flag:rest) =
    case flag of
      "-c" -> expectInt "-c" rest $ \n -> go (cfg {cfgCount = n})
      "--count" -> expectInt "--count" rest $ \n -> go (cfg {cfgCount = n})
      "-f" -> expectValue "-f" rest $ \path -> go (cfg {cfgPath = path})
      "--file" -> expectValue "--file" rest $ \path -> go (cfg {cfgPath = path})
      "--with-replacement" -> go (cfg {cfgWithReplacement = True}) rest
      "--json" -> go (cfg {cfgJSON = True}) rest
      "--stats" -> go (cfg {cfgStats = True}) rest
      "--seed" -> expectInt "--seed" rest $ \n -> go (cfg {cfgSeed = Just n})
      "-h" -> Left usage
      "--help" -> Left usage
      _ -> Left $ "Unknown flag: " ++ flag

  expectInt name xs cont =
    case xs of
      (val:rest) ->
        case safeReadInt val of
          Nothing -> Left $ "Expected integer after " ++ name ++ ", got '" ++ val ++ "'"
          Just n -> cont n rest
      [] -> Left $ "Missing integer after " ++ name

  expectValue name xs cont =
    case xs of
      (val:rest) -> cont val rest
      [] -> Left $ "Missing argument after " ++ name

safeReadInt :: String -> Maybe Int
safeReadInt s = case reads s of
  [(n, "")] -> Just n
  _ -> Nothing

runWithConfig :: Config -> IO ()
runWithConfig cfg@Config {..} = do
  contentExists <- safeReadFile cfgPath
  (names, weights) <- case contentExists of
    Right val -> do
      parsed <- parseNames (lines val)
      let ns = ndNames parsed
      when (null ns) $ do
        hPutStrLn stderr "Name file is empty"
        exitFailure
      pure (ns, ndWeights parsed)
    Left err -> do
      hPutStrLn stderr err
      hPutStrLn stderr "Creating a default names.txt with sample entries."
      writeFile cfgPath (unlines defaultNames)
      pure (defaultNames, Nothing)

  validateConfig cfg names

  gen <- case cfgSeed of
    Just s -> pure (mkStdGen s)
    Nothing -> newStdGen

  let (picks, _) = pickNames gen names weights cfgCount cfgWithReplacement
      statsBlock = if cfgStats then Just (buildStats names weights) else Nothing
  if cfgJSON
    then putStrLn $ renderJSON cfg picks names weights statsBlock
    else do
      putStrLn $ "Picked " ++ show (length picks) ++ " name(s):"
      mapM_ (putStrLn . ("  - " ++)) picks
      maybe (pure ()) (\s -> putStrLn $ "(Deterministic seed " ++ show s ++ ")") cfgSeed
      case statsBlock of
        Nothing -> pure ()
        Just statsJSON -> do
          putStrLn ""
          putStrLn "Statistics:"
          putStrLn statsJSON

-- | Attempt to read a file, returning a Left error message when not possible.
safeReadFile :: FilePath -> IO (Either String String)
safeReadFile path = do
  result <- try (readFile path) :: IO (Either IOException String)
  pure $ case result of
    Right contents -> Right contents
    Left _ -> Left $ "Unable to read " ++ path

-- | Parse the names file, skipping invalid lines with warnings.
parseNames :: [String] -> IO NameData
parseNames = foldM step (NameData [] Nothing)
 where
  step acc rawLine =
    case trim rawLine of
      "" -> pure acc
      line ->
        case break (== ',') line of
          (name, '') -> pure $ appendName acc name 1.0 False
          (name, _:weightStr) ->
            case safeReadDouble (trim weightStr) of
              Just w | w > 0 -> pure $ appendName acc name w True
              _ -> do
                hPutStrLn stderr $ "Warning: invalid weight for '" ++ name ++ "' – skipping line"
                pure acc

  appendName (NameData ns Nothing) name w sawWeight =
    let newWeights = if sawWeight then Just (replicate (length ns) 1.0 ++ [w]) else Nothing
     in NameData (ns ++ [trim name]) newWeights
  appendName (NameData ns (Just ws)) name w _ =
    NameData (ns ++ [trim name]) (Just (ws ++ [w]))

safeReadDouble :: String -> Maybe Double
safeReadDouble s = case reads s of
  [(n, "")] -> Just n
  _ -> Nothing

trim :: String -> String
trim = f . f
 where
  f = reverse . dropWhile isSpace

-- | Validate configuration after parsing input.
validateConfig :: Config -> [String] -> IO ()
validateConfig Config {..} names
  | cfgCount <= 0 = do
      hPutStrLn stderr "Count must be positive"
      exitFailure
  | not cfgWithReplacement && cfgCount > length names = do
      hPutStrLn stderr $ "Cannot pick " ++ show cfgCount ++ " unique names from " ++ show (length names)
      exitFailure
  | otherwise = pure ()

-- | Select names according to the requested replacement mode and weights.
pickNames :: StdGen -> [String] -> Maybe [Double] -> Int -> Bool -> ([String], StdGen)
pickNames gen names weights count withReplacement
  | withReplacement = pickWithReplacement gen count dist []
  | otherwise = pickWithoutReplacement gen count (zip names weightVec) []
 where
  weightVec = fromMaybe (replicate (length names) 1.0) weights
  dist = zip names weightVec

pickWithReplacement :: StdGen -> Int -> [(String, Double)] -> [String] -> ([String], StdGen)
pickWithReplacement gen count dist acc
  | count <= 0 = (reverse acc, gen)
  | null dist = (reverse acc, gen)
  | otherwise =
      let total = sum (map snd dist)
          (r, gen') = randomR (0.0, total) gen
          name = fst (selectWeighted dist r)
       in pickWithReplacement gen' (count - 1) dist (name : acc)

pickWithoutReplacement :: StdGen -> Int -> [(String, Double)] -> [String] -> ([String], StdGen)
pickWithoutReplacement gen count dist acc
  | count <= 0 = (reverse acc, gen)
  | null dist = (reverse acc, gen)
  | otherwise =
      let total = sum (map snd dist)
          (r, gen') = randomR (0.0, total) gen
          (name, remaining) = selectAndRemove dist r
       in pickWithoutReplacement gen' (count - 1) remaining (name : acc)

-- | Select a name by cumulative weight.
selectWeighted :: [(String, Double)] -> Double -> (String, Double)
selectWeighted [] _ = error "selectWeighted called with empty distribution"
selectWeighted ((name, w) : rest) r
  | r <= w = (name, w)
  | null rest = (name, w)
  | otherwise = selectWeighted rest (r - w)

selectAndRemove :: [(String, Double)] -> Double -> (String, [(String, Double)])
selectAndRemove [] _ = error "selectAndRemove called with empty distribution"
selectAndRemove ((name, w) : rest) r
  | r <= w = (name, rest)
  | null rest = (name, rest)
  | otherwise =
      let (picked, remaining) = selectAndRemove rest (r - w)
       in (picked, (name, w) : remaining)

-- | Build a JSON formatted string, optionally embedding statistics.
renderJSON :: Config -> [String] -> [String] -> Maybe [Double] -> Maybe String -> String
renderJSON Config {..} picks names weights statsBlock =
  let baseFields =
        [ jsonField "picked" (jsonArray (map jsonString picks))
        , jsonField "count" (show cfgCount)
        , jsonField "with_replacement" (jsonBool cfgWithReplacement)
        , jsonField "seed" (maybe "null" show cfgSeed)
        , jsonField "total_names" (show (length names))
        , jsonField "weights_used" (jsonBool (maybe False (const True) weights))
        ]
      statsField = maybe [] (\s -> [jsonField "statistics" s]) statsBlock
   in "{" ++ intercalate "," (baseFields ++ statsField) ++ "}"

-- | Construct the statistics block used by the visualiser.
buildStats :: [String] -> Maybe [Double] -> String
buildStats names weights =
  let concreteWeights = fromMaybe (replicate (length names) 1.0) weights
      normalised = normalise concreteWeights
   in
      "{"
        ++ intercalate
          ","
          [ jsonField "names" (jsonArray (map jsonString names))
          , jsonField "weights" (maybe "null" (jsonArray . map jsonNumber) weights)
          , jsonField "normalized_weights" (jsonArray (map jsonNumber normalised))
          ]
        ++ "}"

normalise :: [Double] -> [Double]
normalise ws =
  let total = sum ws
   in if total <= 0 then replicate (length ws) 0 else map (/ total) ws

-- | Render a key-value JSON pair without additional whitespace.
jsonField :: String -> String -> String
jsonField key value = jsonString key ++ ":" ++ value

jsonArray :: [String] -> String
jsonArray xs = "[" ++ intercalate "," xs ++ "]"

jsonBool :: Bool -> String
jsonBool True = "true"
jsonBool False = "false"

jsonString :: String -> String
jsonString s = '"' : concatMap escapeChar s ++ ""
 where
  escapeChar '"' = "\\\""
  escapeChar '\\' = "\\\\"
  escapeChar '\n' = "\\n"
  escapeChar '\r' = "\\r"
  escapeChar '\t' = "\\t"
  escapeChar c
    | c < ' ' = "\\u" ++ hex4 (fromEnum c)
    | otherwise = [c]
  hex4 n = let hex = "0123456789abcdef" in [hex !! (n `div` 4096), hex !! ((n `div` 256) `mod` 16), hex !! ((n `div` 16) `mod` 16), hex !! (n `mod` 16)]

jsonNumber :: Double -> String
jsonNumber = trimZeroes . show
 where
  trimZeroes str =
    case break (== '.') str of
      (intPart, '.' : frac) ->
        let frac' = reverse (dropWhile (== '0') (reverse frac))
         in if null frac' then intPart else intPart ++ "." ++ frac'
      _ -> str
